src/Microsoft.ML.StandardLearners/NearestNeighbor/MultiClassNearestNeighborTrainer.cs

```diff
@@ -0,0 +1,273 @@
+        /// </summary>
+        internal MultiClassNearestNeighborTrainer(IHostEnvironment env, Arguments args)
+            : base(env, args, TrainerUtils.MakeU4ScalarColumn(args.LabelColumn))
+        {
+        }
+
+        protected override SchemaShape.Column[] GetOutputColumnsCore(SchemaShape inputSchema)
+        {
+            bool success = inputSchema.TryFindColumn(LabelColumn.Name, out var labelCol);
+            Contracts.Assert(success);
+
+            var predLabelMetadata = new SchemaShape(labelCol.Metadata.Where(x => x.Name == MetadataUtils.Kinds.KeyValues)
+                .Concat(MetadataUtils.GetTrainerOutputMetadata()));
+
+            return new[]
+            {
+                new SchemaShape.Column(DefaultColumnNames.Score, SchemaShape.Column.VectorKind.Vector, NumberType.R4, false, new SchemaShape(MetadataUtils.MetadataForMulticlassScoreColumn(labelCol))),
+                new SchemaShape.Column(DefaultColumnNames.PredictedLabel, SchemaShape.Column.VectorKind.Scalar, NumberType.U4, true, predLabelMetadata)
+            };
+        }
+
+        protected override MulticlassPredictionTransformer<MultiClassNearestNeighborModelParameters> MakeTransformer(MultiClassNearestNeighborModelParameters model, Schema trainSchema)
+            => new MulticlassPredictionTransformer<MultiClassNearestNeighborModelParameters>(Host, model, trainSchema, FeatureColumn.Name, LabelColumn.Name);
+
+        public MulticlassPredictionTransformer<MultiClassNearestNeighborModelParameters> Train(IDataView trainData, IPredictor initialPredictor = null)
+            => TrainTransformer(trainData, initPredictor: initialPredictor);
+
+        private protected override MultiClassNearestNeighborModelParameters TrainModelCore(TrainContext context)
+        {
+            Host.CheckValue(context, nameof(context));
+            var data = context.TrainingSet;
+            Host.Check(data.Schema.Label != null, "Missing Label column");
+            Host.Check(data.Schema.Label.Type == NumberType.Float || data.Schema.Label.Type is KeyType,
+                "Invalid type for Label column, only floats and known-size keys are supported");
+
+            Host.Check(data.Schema.Feature != null, "Missing Feature column");
+            int featureCount;
+            data.CheckFeatureFloatVector(out featureCount);
+            int labelCount = 0;
+            if (data.Schema.Label.Type.IsKey)
+                labelCount = data.Schema.Label.Type.KeyCount;
+
+            var examples = new List<VBuffer<float>>();
+            var labels = new List<int>();
+            if (context.InitialPredictor is MultiClassNearestNeighborModelParameters oldPredictor)
+            {
+                examples.AddRange(oldPredictor.Objects);
+                labels.AddRange(oldPredictor.Labels);
+            }
+            using (var ch = Host.Start("Data extraction"))
+            using (var cursor = new MultiClassLabelCursor(labelCount, data, CursOpt.Features | CursOpt.Label))
```

Comment:
>using [](start = 12, length = 5)

I wonder if the entire training loop can be replaced with 'choose columns transformer' on top of training data to limit it to just features and label, and then saving the data view?